### ISSUE TITLE
Clang ignores function templates introduced by using declaration during overload resolution with explicit non-type template arguments

### ISSUE THREAD

**Issue by Narkodav** [llvm/llvm-project#206198]
# Function templates introduced via `using` declaration are ignored during overload resolution with explicit non-type template arguments

## Compiler Version

```
clang version 22.1.0-rc3
Target: x86_64-w64-windows-gnu
Thread model: posix
```

The issue is also reproducible on all Clang versions currently available on Compiler Explorer.
### Compiler Explorer reproducer:
https://godbolt.org/z/4vG1qz6oW

## Minimal reproducer

```cpp
#include <iostream>

struct Base {
    template<int N>
    void f() {
        std::cout << "Base\n";
    }
};

struct Derived : Base {
    using Base::f;

    template<char C>
    void f() {
        std::cout << "Derived\n";
    }
};

int main() {
    Derived d;
    d.f<1>();
}
```

## Expected behavior

The `using Base::f;` declaration introduces both function templates into the overload set.

The explicit template argument `1` is a valid converted constant expression for both `int` and `char`, making both templates viable candidates.

After substitution, both candidates have the same signature:

```cpp
void f();
```

Neither template is more specialized than the other, so overload resolution should report the call as ambiguous.

Both GCC and MSVC reject the program with an ambiguity diagnostic.

## Actual behavior

Clang accepts the program and calls `Derived::f`, apparently ignoring the inherited template introduced by the `using` declaration during overload resolution.

## Related reproducer

The same issue appears when the non-type template parameters have unrelated enum types.

```cpp
#include <iostream>

enum class E1 { A };
enum class E2 { B };

struct Base {
    template<E1 E>
    void f() {
        std::cout << "Base\n";
    }
};

struct Derived : Base {
    using Base::f;

    template<E2 E>
    void f() {
        std::cout << "Derived\n";
    }
};

int main() {
    Derived d;
    d.f<E1::A>();
}
```

### Expected behavior

The call should invoke `Base::f`, since only the base template is viable.

GCC and MSVC both compile this program and print:

```
Base
```

### Actual behavior

Clang rejects the program with:

```
error: no matching member function for call to 'f'
note: candidate template ignored: invalid explicitly-specified argument
```

The diagnostic only lists the derived template as a candidate, suggesting that the inherited template introduced by the `using` declaration is not considered.

## Practical impact

I encountered this issue while implementing an event framework that uses inheritance to extend the set of supported events.

A simplified version looks like this:

```cpp
struct Base {
    template<BaseEvent E>
    void subscribe(...);
};

struct Derived : Base {
    using Base::subscribe;

    template<DerivedEvent E>
    void subscribe(...);
};
```

The intended API is:

```cpp
Derived d;

d.subscribe<BaseEvent::Foo>(...);
d.subscribe<DerivedEvent::Bar>(...);
```

This works correctly with GCC and MSVC, but with Clang only the derived overloads are accessible unless the base class is explicitly qualified.

## Compiler comparison

| Compiler | First reproducer   | Second reproducer                                 |
| -------- | ------------------ | ------------------------------------------------- |
| GCC      | Ambiguous          | Calls `Base::f`                                   |
| MSVC     | Ambiguous          | Calls `Base::f`                                   |
| Clang    | Calls `Derived::f` | Rejects program; only derived template considered |

**Comment by EugeneZelenko** [comment]
@Narkodav: Please use GitHub markup instead of HTML. Will be good idea to provide reproducer on https://godbolt.org.

**Comment by Narkodav** [comment]
Sure, changed to GitHub markup and added a Compiler Explorer reproducer

**Comment by zygoloid** [comment]
This is a change from [P1787R6](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1787r6.html):

Old rule:
> When a using-declarator brings declarations from a base class into a derived class, member functions and member function templates in the derived class override and/or hide member functions and member function templates with the same name, parameter-type-list ([dcl.fct]), trailing requires-clause (if any), cv-qualification, and ref-qualifier (if any), in a base class (rather than conflicting). Such hidden or overridden declarations are excluded from the set of declarations introduced by the using-declarator. [Example:

New rule:
> The set of declarations named by a using-declarator that inhabits a class C does not include member functions and member function templates of a base class that correspond to (and thus would conflict with) a declaration of a function or function template in C. [Example:

Note that the old rule only considers the name, parameter types, requires clause, and cv / ref qualifiers. The new rule uses a "correspond to", which is a full redeclaration check, including checking template parameters.

Clang still implements the old rule; the other compilers seem to implement the new rule.

**Comment by shafik** [comment]
There is a bit of fluff in this report that could be avoided. We don't need to repeat the title in the summary, we don't need the compiler version, target or thread model unless those specific attributes are required to reproduce. Although if this behavior starts w/ a specific version of clang that often is helpful to know.
